<a href="https://colab.research.google.com/github/jocinenabwire/WEEK-3-IP/blob/main/Jocine_Nabwire_Python_Notebook_Week_3_IP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MORINGA PREP DATA SCIENCE INDEPENDENT PROJECT: WEEK 2

In this week's independent project, you will be working as Data Scientist for MTN Cote d'Ivoire, a leading telecom company and you will be solving for the following research question.

Currently MTN Cote d'Ivoire would like to upgrade its technology infrastructure for its mobile users in Ivory Coast. Studying the given dataset, how does MTN Cote d'Ivoire go about the upgrade of its infrastructure strategy within the given cities?

Below are some questions that can get you started;

1.Which ones were the most used city for the three days?

2.Which cities were the most used during business and home hours?

3.Most used city for the three days?

etc. 

#**Objective**

Finding how MTN Cote d'Ivoire will go about the upgrade of its infrastructure strategy within the given cities?

**Dataset Provided**

1.cells_geo_description.xlsx [Link]

2.cells_geo.csv [Link]

3.CDR_description.xlsx [Link]

4.CDR 20120507 [http://bit.ly/TelecomDataset1]

5.CDR 20120508 [http://bit.ly/TelecomDataset2]

6.CDR 20120509 [http://bit.ly/TelecomDataset3]

#Importing libraries

In [609]:
#we first import the pandas library
import pandas as pd
import numpy as np

#Loading and viewing Datasets 

After importing and loading our data sets, we can view the DataFrames

In [610]:
#viweing our Cells Geo csv
url= '/content/cells_geo.csv'
df_cells = pd.read_csv(url, delimiter=';')
df_cells

Unnamed: 0    VILLES      STATUS  ...          AREA     CELL_ID   SITE_CODE
0              0    ADJAME  In Service  ...           NaN  fbb2e75af3  0bb492ef17
1              1    ADJAME  In Service  ...           NaN  1502501d78  0bb492ef17
2              2    ADJAME  In Service  ...           NaN  0c2e182a0c  0bb492ef17
3              3   ASSINIE  In Service  ...           NaN  69bdfc69fd  8a1e6ce31f
4              4   ASSINIE  In Service  ...           NaN  7a1029d2ab  8a1e6ce31f
...          ...       ...         ...  ...           ...         ...         ...
3969        3969  ZUENOULA  In Service  ...  CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970        3970  ZUENOULA  In Service  ...  CENTRE NORTH  3ae1708153  f91ed8d2e9
3971        3971  ZUENOULA  In Service  ...  CENTRE NORTH  cb71f8a29c  61bee9c70c
3972        3972  ZUENOULA  In Service  ...  CENTRE NORTH  d4a6a0b645  61bee9c70c
3973        3973  ZUENOULA  In Service  ...  CENTRE NORTH  f9f3e42589  61bee9c70c

[3974 rows x 12 columns]

In [611]:
#we rename some of the columns to English 
df_cells=df_cells.rename(columns={'VILLES':'CITY','LOCALISATION':'LOCATION','DECOUPZONE':'GEOGRAPHICAL_ZONE'})
df_cells

Unnamed: 0      CITY      STATUS  ...          AREA     CELL_ID   SITE_CODE
0              0    ADJAME  In Service  ...           NaN  fbb2e75af3  0bb492ef17
1              1    ADJAME  In Service  ...           NaN  1502501d78  0bb492ef17
2              2    ADJAME  In Service  ...           NaN  0c2e182a0c  0bb492ef17
3              3   ASSINIE  In Service  ...           NaN  69bdfc69fd  8a1e6ce31f
4              4   ASSINIE  In Service  ...           NaN  7a1029d2ab  8a1e6ce31f
...          ...       ...         ...  ...           ...         ...         ...
3969        3969  ZUENOULA  In Service  ...  CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970        3970  ZUENOULA  In Service  ...  CENTRE NORTH  3ae1708153  f91ed8d2e9
3971        3971  ZUENOULA  In Service  ...  CENTRE NORTH  cb71f8a29c  61bee9c70c
3972        3972  ZUENOULA  In Service  ...  CENTRE NORTH  d4a6a0b645  61bee9c70c
3973        3973  ZUENOULA  In Service  ...  CENTRE NORTH  f9f3e42589  61bee9c70c

[3974 rows x 12 columns]

In [612]:
#viweing our Telcom Dataset(CDR 20120507)
df_20120507 = pd.read_csv('/content/Telcom_dataset.csv')
df_20120507 

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2        sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997   Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000   Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b77a5  ebfc58af70

[5001 rows x 10 columns]

In [613]:
#viweing our Telcom Dataset 2(CDR 20120508)
df_20120508 = pd.read_csv('/content/Telcom_dataset2.csv')
df_20120508

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0        sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1      Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2        sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3      Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4        sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms      0  2012-05-08 00:01:53.0  ...     19e2e  896459aace  b7d00e59e4
4997     sms     25  2012-05-08 00:01:53.0  ...     19e2e  7e90b35166  211015b14d
4998     sms      0  2012-05-08 00:01:53.0  ...     19e2e  fe70e431df  c7ea989264
4999    data      0  2012-05-08 00:01:53.0  ...     OTHER  ffa6759bb2         NaN
5000   Voice      0  2012-05-08 00:01:53.0  ...     19e2e  8630058aca  b47a998a3e

[5001 rows x 10 columns]

In [614]:
#viweing our Telcom Dataset 3(CDR 20120509)
df_20120509= pd.read_csv('/content/Telcom_dataset3.csv')
df_20120509

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0      Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1        sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2      Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3        sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA
...      ...    ...                    ...  ...       ...         ...         ...
4996   Voice      0  2012-05-09 00:01:43.0  ...     19e2e  2aa04dcab7  f7d63a29e5
4997   Voice     62  2012-05-09 00:01:43.0  ...     19e2e  7f1f2f889f  b4ddb9e93c
4998   Voice     32  2012-05-09 00:01:43.0  ...     19e2e  a4731e780e  d68260ef50
4999   Voice     30  2012-05-09 00:01:43.0  ...     19e2e  98ed726da2  b4130b54ab
5000   Voice     60  2012-05-09 00:01:43.0  ...     19e2e  1c11a2fcbb  6d70dae094

[5001 rows x 10 columns]

#**Data Understanding**

Cells Data


In [615]:
#Describing the columns in the Cells data
df_cells.columns

Index(['Unnamed: 0', 'CITY', 'STATUS', 'LOCATION', 'GEOGRAPHICAL_ZONE',
       'ZONENAME', 'LONGITUDE', 'LATITUDE', 'REGION', 'AREA', 'CELL_ID',
       'SITE_CODE'],
      dtype='object')

There are twelve columns in the in the Cells DataFrame.

In [616]:
#Getting information on the DataFrame
df_cells.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3974 entries, 0 to 3973
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         3974 non-null   int64  
 1   CITY               3974 non-null   object 
 2   STATUS             3907 non-null   object 
 3   LOCATION           3974 non-null   object 
 4   GEOGRAPHICAL_ZONE  3974 non-null   object 
 5   ZONENAME           3968 non-null   object 
 6   LONGITUDE          3974 non-null   float64
 7   LATITUDE           3974 non-null   float64
 8   REGION             3974 non-null   object 
 9   AREA               3951 non-null   object 
 10  CELL_ID            3974 non-null   object 
 11  SITE_CODE          3974 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 372.7+ KB


Two columns have float data type, one column has integer data type and nine columns have object data type.

In [617]:
#Getting the null values
df_cells.isnull().sum()

Unnamed: 0            0
CITY                  0
STATUS               67
LOCATION              0
GEOGRAPHICAL_ZONE     0
ZONENAME              6
LONGITUDE             0
LATITUDE              0
REGION                0
AREA                 23
CELL_ID               0
SITE_CODE             0
dtype: int64

Status column has 67 null values. Area column has 23 null values.

In [618]:
#Getting the shape of the DataFrame
df_cells.shape


(3974, 12)

The Cells DataFrame has 3974 rows and 12 columns.

In [619]:
#Describing the Cells DataFrame
df_cells.describe()

Unnamed: 0    LONGITUDE     LATITUDE
count  3974.000000  3974.000000  3974.000000
mean   1986.500000    -4.877659     6.244749
std    1147.339313     1.300771     1.304199
min       0.000000    -8.503889     0.000000
25%     993.250000    -5.846139     5.338639
50%    1986.500000    -4.084500     5.475000
75%    2979.750000    -3.993444     6.872306
max    3973.000000    -0.000000    10.481861

The count, mean,std,etc is as shown above.

Telcom DataSet

a)df_20120507

In [620]:
#Describing the columns
df_20120507.columns

Index(['PRODUTC', 'VALUE', 'DATETIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELL_ID', 'SITE_ID'],
      dtype='object')

In [621]:
#Getting info
df_20120507.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUTC          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATETIME         5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELL_ID          5001 non-null   object
 9   SITE_ID          4285 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [622]:
#Getting the null values
df_20120507.isnull().sum()

PRODUTC              0
VALUE                0
DATETIME             0
CELL_ON_SITE         0
DW_A_NUMBER_INT      0
DW_B_NUMBER_INT      0
COUNTRY_A            0
COUNTRY_B            0
CELL_ID              0
SITE_ID            716
dtype: int64

Site ID column is the only one with null values

In [623]:
#Getting the shape of the DataFrame
df_20120507.shape


(5001, 10)

In [624]:
#Describing the DataFrame
df_20120507.describe()

VALUE
count  5001.000000
mean     25.869626
std     101.934502
min       0.000000
25%       0.000000
50%       0.000000
75%      25.000000
max    4440.000000

In [625]:
df_20120507.nunique()

PRODUTC               3
VALUE               242
DATETIME            708
CELL_ON_SITE          7
DW_A_NUMBER_INT    3903
DW_B_NUMBER_INT    2848
COUNTRY_A             1
COUNTRY_B             2
CELL_ID            1790
SITE_ID            1076
dtype: int64

Shows the unique number of unique values in each column

In [626]:
#Getting specific unique values in each column
df_20120507.PRODUTC.unique()

array(['Voice', 'sms', 'data'], dtype=object)

Product column has three unique values,i.e, voice,sms and data.

In [627]:
df_20120507.CELL_ON_SITE.unique()

array(['1', '0', '_', '5', '2', '3', '4'], dtype=object)

In [628]:
df_20120507.COUNTRY_A.unique()

array(['19e2e'], dtype=object)

In [629]:
df_20120507.COUNTRY_B.unique()

array(['19e2e', 'OTHER'], dtype=object)

b)df_20120508 

In [630]:
#Describing the columns
df_20120508.columns

Index(['PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'DW_A_NUMBER',
       'DW_B_NUMBER', 'COUNTRY_A', 'COUNTRY_B', 'CELL_ID', 'SITE_ID'],
      dtype='object')

In [631]:
#Getting info
df_20120508.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PRODUCT       5001 non-null   object
 1   VALUE         5001 non-null   int64 
 2   DATE_TIME     5001 non-null   object
 3   CELL_ON_SITE  5001 non-null   object
 4   DW_A_NUMBER   5001 non-null   object
 5   DW_B_NUMBER   5001 non-null   object
 6   COUNTRY_A     5001 non-null   object
 7   COUNTRY_B     5001 non-null   object
 8   CELL_ID       5001 non-null   object
 9   SITE_ID       4325 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [632]:
#Getting the null values
df_20120508.isnull().sum()

PRODUCT           0
VALUE             0
DATE_TIME         0
CELL_ON_SITE      0
DW_A_NUMBER       0
DW_B_NUMBER       0
COUNTRY_A         0
COUNTRY_B         0
CELL_ID           0
SITE_ID         676
dtype: int64

In [633]:
#Getting the shape of the DataFrame
df_20120508.shape

(5001, 10)

In [634]:
#Describing the DataFrame
df_20120508.describe()

VALUE
count  5001.000000
mean     24.334933
std      97.471586
min       0.000000
25%       0.000000
50%       0.000000
75%      25.000000
max    3380.000000

In [635]:
df_20120508.nunique()

PRODUCT            3
VALUE            242
DATE_TIME        788
CELL_ON_SITE       8
DW_A_NUMBER     3799
DW_B_NUMBER     2882
COUNTRY_A          1
COUNTRY_B          2
CELL_ID         1661
SITE_ID         1027
dtype: int64

In [636]:
#Getting specific unique values in each column
df_20120508.PRODUCT.unique()

array(['sms', 'Voice', 'data'], dtype=object)

Product column has three unique values,i.e, voice,sms and data.

In [637]:
df_20120508.CELL_ON_SITE.unique()

array(['2', '1', '0', '4', '_', '3', '5', '8'], dtype=object)

In [638]:
df_20120508.COUNTRY_A.unique()

array(['19e2e'], dtype=object)

In [639]:
df_20120508.COUNTRY_B.unique()

array(['19e2e', 'OTHER'], dtype=object)

c)df_20120509

In [640]:
#Describing the columns
df_20120509.columns

Index(['PRODUCT', 'VALUE', 'DATE_TIME', 'CELL_ON_SITE', 'DW_A_NUMBER_INT',
       'DW_B_NUMBER_INT', 'COUNTRY_A', 'COUNTRY_B', 'CELLID', 'SIET_ID'],
      dtype='object')

In [641]:
#Getting info
df_20120509.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUCT          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATE_TIME        5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELLID           5001 non-null   object
 9   SIET_ID          4396 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB


In [642]:
#Getting the null values
df_20120509.isnull().sum()

PRODUCT              0
VALUE                0
DATE_TIME            0
CELL_ON_SITE         0
DW_A_NUMBER_INT      0
DW_B_NUMBER_INT      0
COUNTRY_A            0
COUNTRY_B            0
CELLID               0
SIET_ID            605
dtype: int64

In [643]:
#Getting the shape of the DataFrame
df_20120509.shape

(5001, 10)

In [644]:
#Describing the DataFrame
df_20120509.describe()

VALUE
count   5001.000000
mean      26.441512
std      224.200217
min        0.000000
25%        0.000000
50%        0.000000
75%       25.000000
max    12900.000000

In [645]:
df_20120509.nunique()

PRODUCT               3
VALUE               219
DATE_TIME           647
CELL_ON_SITE          9
DW_A_NUMBER_INT    3593
DW_B_NUMBER_INT    2895
COUNTRY_A             1
COUNTRY_B             2
CELLID             1596
SIET_ID             961
dtype: int64

In [646]:
#Getting specific unique values in each column
df_20120509.PRODUCT.unique()

array(['Voice', 'sms', 'data'], dtype=object)

Product column has three unique values,i.e, voice,sms and data.

In [647]:
df_20120509.CELL_ON_SITE.unique()

array(['2', '1', '0', '4', '_', '5', '3', '6', '7'], dtype=object)

In [648]:
df_20120509.COUNTRY_A.unique()

array(['19e2e'], dtype=object)

In [649]:
df_20120509.COUNTRY_B.unique()

array(['19e2e', 'OTHER'], dtype=object)

In [650]:
#Checking for duplicates
df_20120507[df_20120507.duplicated(keep=False)]

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
765      sms      0  2012-05-07 00:00:03.0  ...     19e2e  23c5372bd3  e942f32a29
769      sms     25  2012-05-07 00:00:03.0  ...     19e2e  a6606f4ac1  c6400e443f
793      sms      0  2012-05-07 00:00:03.0  ...     19e2e  23c5372bd3  e942f32a29
794      sms     25  2012-05-07 00:00:03.0  ...     19e2e  a6606f4ac1  c6400e443f
795    Voice      0  2012-05-07 00:00:04.0  ...     19e2e  4a7047f1ba  dea1e52d3b
...      ...    ...                    ...  ...       ...         ...         ...
4864   Voice   1610  2012-05-07 00:01:50.0  ...     19e2e  89b82605a0  5324bc6c24
4932   Voice      0  2012-05-07 00:01:52.0  ...     19e2e  d6abb7e15f  f72cd95f00
4943   Voice      0  2012-05-07 00:01:52.0  ...     19e2e  d6abb7e15f  f72cd95f00
4944   Voice     60  2012-05-07 00:01:53.0  ...     19e2e  fd754bd0d0  e82c07d996
4970   Voice     60  2012-05-07 00:01:53.0  ...     19e2e  fd754bd0d0  e82c07d996

[164 rows x 10 columns]

There are duplicates found.

In [651]:
df_20120508[df_20120508.duplicated(keep=False)]

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
801    Voice     42  2012-05-08 00:00:01.0  ...     19e2e  7a31902854  55d6d9039c
823    Voice     42  2012-05-08 00:00:01.0  ...     19e2e  7a31902854  55d6d9039c
824    Voice     46  2012-05-08 00:00:02.0  ...     19e2e  4d9b2590ad  0c997709bd
825    Voice      0  2012-05-08 00:00:02.0  ...     19e2e  3aba7c72d7  e188165ebc
836    Voice     46  2012-05-08 00:00:02.0  ...     19e2e  4d9b2590ad  0c997709bd
...      ...    ...                    ...  ...       ...         ...         ...
4879     sms     25  2012-05-08 00:01:50.0  ...     19e2e  777d30093d  aeed65e7ce
4880     sms      0  2012-05-08 00:01:51.0  ...     19e2e  15bdb525ee  6e9b544976
4918     sms      0  2012-05-08 00:01:51.0  ...     19e2e  15bdb525ee  6e9b544976
4973   Voice      0  2012-05-08 00:01:53.0  ...     19e2e  8630058aca  b47a998a3e
5000   Voice      0  2012-05-08 00:01:53.0  ...     19e2e  8630058aca  b47a998a3e

[156 rows x 10 columns]

There are duplicates found.

In [652]:
df_20120509[df_20120509.duplicated(keep=False)]

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
833      sms     25  2012-05-09 00:00:05.0  ...     19e2e  bf65ca9bc0  ba8ce6e4ef
879     data      0  2012-05-09 00:00:05.0  ...     OTHER  ffa6759bb2         NaN
892      sms     25  2012-05-09 00:00:05.0  ...     19e2e  bf65ca9bc0  ba8ce6e4ef
893     data      0  2012-05-09 00:00:05.0  ...     OTHER  ffa6759bb2         NaN
894      sms     25  2012-05-09 00:00:06.0  ...     19e2e  07ed226112  334fdb9912
...      ...    ...                    ...  ...       ...         ...         ...
4962     sms      0  2012-05-09 00:01:42.0  ...     19e2e  34f124cb0d  73dd78713b
4963     sms      0  2012-05-09 00:01:43.0  ...     19e2e  55becafe68  7430461f18
4964   Voice     32  2012-05-09 00:01:43.0  ...     19e2e  a4731e780e  d68260ef50
4969     sms      0  2012-05-09 00:01:43.0  ...     19e2e  55becafe68  7430461f18
4998   Voice     32  2012-05-09 00:01:43.0  ...     19e2e  a4731e780e  d68260ef50

[186 rows x 10 columns]

There are duplicates found.

#**Data Cleaning**


We have to rename some of the column names in the Telcom Datasets for uniformity

In [653]:
#df_20120507
#Renaming columns DATETIME
df_20120507=df_20120507.rename(columns={'PRODUTC':'PRODUCT','DATETIME':'DATE_TIME'})
df_20120507

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2        sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms    100  2012-05-07 00:01:54.0  ...     OTHER  0c2371d9bc  1e751d08a4
4997   Voice      1  2012-05-07 00:01:54.0  ...     19e2e  5fb791cf56  65efe2bc62
4998   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  7773627c12  9ef9616350
4999   Voice      0  2012-05-07 00:01:54.0  ...     19e2e  3f4af49d6c  969b8d0e39
5000   Voice      7  2012-05-07 00:01:54.0  ...     19e2e  6f4d9b77a5  ebfc58af70

[5001 rows x 10 columns]

In [654]:
#df_20120508
#Renaming columns DW_A_NUMBER and DW_B_NUMBER
df_20120508=df_20120508.rename(columns={'DW_A_NUMBER':'DW_A_NUMBER_INT','DW_B_NUMBER':'DW_B_NUMBER_INT'})
df_20120508

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0        sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1      Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2        sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3      Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4        sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO
...      ...    ...                    ...  ...       ...         ...         ...
4996     sms      0  2012-05-08 00:01:53.0  ...     19e2e  896459aace  b7d00e59e4
4997     sms     25  2012-05-08 00:01:53.0  ...     19e2e  7e90b35166  211015b14d
4998     sms      0  2012-05-08 00:01:53.0  ...     19e2e  fe70e431df  c7ea989264
4999    data      0  2012-05-08 00:01:53.0  ...     OTHER  ffa6759bb2         NaN
5000   Voice      0  2012-05-08 00:01:53.0  ...     19e2e  8630058aca  b47a998a3e

[5001 rows x 10 columns]

In [655]:
#df_20120509
#Renaming columns SIET_ID and CELLID
df_20120509=df_20120509.rename(columns={'CELLID':'CELL_ID','SIET_ID':'SITE_ID'})
df_20120509

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0      Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1        sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2      Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3        sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4      Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA
...      ...    ...                    ...  ...       ...         ...         ...
4996   Voice      0  2012-05-09 00:01:43.0  ...     19e2e  2aa04dcab7  f7d63a29e5
4997   Voice     62  2012-05-09 00:01:43.0  ...     19e2e  7f1f2f889f  b4ddb9e93c
4998   Voice     32  2012-05-09 00:01:43.0  ...     19e2e  a4731e780e  d68260ef50
4999   Voice     30  2012-05-09 00:01:43.0  ...     19e2e  98ed726da2  b4130b54ab
5000   Voice     60  2012-05-09 00:01:43.0  ...     19e2e  1c11a2fcbb  6d70dae094

[5001 rows x 10 columns]

In [670]:
#changing the date column to date time data type
df_Telcom['DATE_TIME']= pd.to_datetime(df_Telcom['DATE_TIME'])
df_Telcom['DATE_TIME']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


0      2012-05-06 23:04:37
1      2012-05-06 23:05:25
2      2012-05-06 23:05:41
3      2012-05-06 23:05:42
4      2012-05-06 23:06:29
               ...        
4995   2012-05-09 00:01:43
4996   2012-05-09 00:01:43
4997   2012-05-09 00:01:43
4999   2012-05-09 00:01:43
5000   2012-05-09 00:01:43
Name: DATE_TIME, Length: 14750, dtype: datetime64[ns]

In [657]:
df_Telcom.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14750 entries, 0 to 5000
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   PRODUTC          4919 non-null   object        
 1   VALUE            14750 non-null  int64         
 2   DATE_TIME        14750 non-null  datetime64[ns]
 3   CELL_ON_SITE     14750 non-null  object        
 4   DW_A_NUMBER_INT  14750 non-null  object        
 5   DW_B_NUMBER_INT  14750 non-null  object        
 6   CELL_ID          14750 non-null  object        
 7   SITE_ID          12789 non-null  object        
 8   PRODUCT          9831 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(7)
memory usage: 1.1+ MB


Since we have some columns present in more than one file, we will have to concatenate the DataFrames.

In [658]:
#we use the concat function
df_Telcom=pd.concat([df_20120507,df_20120508,df_20120509], axis=0)
df_Telcom.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [659]:
#getting info
df_Telcom.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15003 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUCT          15003 non-null  object
 1   VALUE            15003 non-null  int64 
 2   DATE_TIME        15003 non-null  object
 3   CELL_ON_SITE     15003 non-null  object
 4   DW_A_NUMBER_INT  15003 non-null  object
 5   DW_B_NUMBER_INT  15003 non-null  object
 6   COUNTRY_A        15003 non-null  object
 7   COUNTRY_B        15003 non-null  object
 8   CELL_ID          15003 non-null  object
 9   SITE_ID          13006 non-null  object
dtypes: int64(1), object(9)
memory usage: 1.3+ MB


In [660]:
#we then drop columns that we don't need
df_Telcom=df_Telcom.drop(['COUNTRY_A','COUNTRY_B'],axis=1)
df_Telcom

PRODUCT  VALUE  ...     CELL_ID     SITE_ID
0      Voice      0  ...  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  ...  /+cKZKIp41  /+cKZKIp41
2        sms     25  ...  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  ...  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  ...  051KhYDCpv  051KhYDCpv
...      ...    ...  ...         ...         ...
4996   Voice      0  ...  2aa04dcab7  f7d63a29e5
4997   Voice     62  ...  7f1f2f889f  b4ddb9e93c
4998   Voice     32  ...  a4731e780e  d68260ef50
4999   Voice     30  ...  98ed726da2  b4130b54ab
5000   Voice     60  ...  1c11a2fcbb  6d70dae094

[15003 rows x 8 columns]

In [661]:
#we then drop the duplicates
df_Telcom = df_Telcom.drop_duplicates()
df_Telcom 

PRODUCT  VALUE  ...     CELL_ID     SITE_ID
0      Voice      0  ...  /TJNe+Mmtu  /TJNe+Mmtu
1        sms      0  ...  /+cKZKIp41  /+cKZKIp41
2        sms     25  ...  /u0FSD+ahi  /u0FSD+ahi
3      Voice     56  ...  +cNeJzsTp3  +cNeJzsTp3
4      Voice     14  ...  051KhYDCpv  051KhYDCpv
...      ...    ...  ...         ...         ...
4995   Voice     10  ...  6467daa1db  8a96bae9c1
4996   Voice      0  ...  2aa04dcab7  f7d63a29e5
4997   Voice     62  ...  7f1f2f889f  b4ddb9e93c
4999   Voice     30  ...  98ed726da2  b4130b54ab
5000   Voice     60  ...  1c11a2fcbb  6d70dae094

[14750 rows x 8 columns]

In [662]:
#Getting the info after dropping the duplicates
df_Telcom.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14750 entries, 0 to 5000
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUCT          14750 non-null  object
 1   VALUE            14750 non-null  int64 
 2   DATE_TIME        14750 non-null  object
 3   CELL_ON_SITE     14750 non-null  object
 4   DW_A_NUMBER_INT  14750 non-null  object
 5   DW_B_NUMBER_INT  14750 non-null  object
 6   CELL_ID          14750 non-null  object
 7   SITE_ID          12789 non-null  object
dtypes: int64(1), object(7)
memory usage: 1.0+ MB


Previously we had 15003 rows. After dropping the duplicates, we remain with 14750 rows. This shows that there were 253 duplicate rows.

In [663]:
#we can then merge the two DataFrames,i.e, df_Cells and df_Telcom using two methods.
#We merge them  using the CELL_ID column which is present in both DataFrames.
#method 1:merge()
df_cells_Telcom=pd.merge(df_Telcom,df_cells,how='inner',on='CELL_ID')
df_cells_Telcom

Empty DataFrame
Columns: [PRODUCT, VALUE, DATE_TIME, CELL_ON_SITE, DW_A_NUMBER_INT, DW_B_NUMBER_INT, CELL_ID, SITE_ID, Unnamed: 0, CITY, STATUS, LOCATION, GEOGRAPHICAL_ZONE, ZONENAME, LONGITUDE, LATITUDE, REGION, AREA, SITE_CODE]
Index: []

In [664]:
#Method 2:concat()
df_cells_Telcom=pd.concat([df_Telcom,df_cells,],axis=0)
df_cells_Telcom

PRODUCT  VALUE              DATE_TIME  ...    REGION          AREA   SITE_CODE
0      Voice    0.0  2012-05-06 23:04:37.0  ...       NaN           NaN         NaN
1        sms    0.0  2012-05-06 23:05:25.0  ...       NaN           NaN         NaN
2        sms   25.0  2012-05-06 23:05:41.0  ...       NaN           NaN         NaN
3      Voice   56.0  2012-05-06 23:05:42.0  ...       NaN           NaN         NaN
4      Voice   14.0  2012-05-06 23:06:29.0  ...       NaN           NaN         NaN
...      ...    ...                    ...  ...       ...           ...         ...
3969     NaN    NaN                    NaN  ...  MARAHOUE  CENTRE NORTH  f91ed8d2e9
3970     NaN    NaN                    NaN  ...  MARAHOUE  CENTRE NORTH  f91ed8d2e9
3971     NaN    NaN                    NaN  ...  MARAHOUE  CENTRE NORTH  61bee9c70c
3972     NaN    NaN                    NaN  ...  MARAHOUE  CENTRE NORTH  61bee9c70c
3973     NaN    NaN                    NaN  ...  MARAHOUE  CENTRE NORTH  61bee9c70c

[18724 rows x 19 columns]

In [665]:
#Getting the info after merging the data frames
df_cells_Telcom.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18724 entries, 0 to 3973
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   PRODUCT            14750 non-null  object 
 1   VALUE              14750 non-null  float64
 2   DATE_TIME          14750 non-null  object 
 3   CELL_ON_SITE       14750 non-null  object 
 4   DW_A_NUMBER_INT    14750 non-null  object 
 5   DW_B_NUMBER_INT    14750 non-null  object 
 6   CELL_ID            18724 non-null  object 
 7   SITE_ID            12789 non-null  object 
 8   Unnamed: 0         3974 non-null   float64
 9   CITY               3974 non-null   object 
 10  STATUS             3907 non-null   object 
 11  LOCATION           3974 non-null   object 
 12  GEOGRAPHICAL_ZONE  3974 non-null   object 
 13  ZONENAME           3968 non-null   object 
 14  LONGITUDE          3974 non-null   float64
 15  LATITUDE           3974 non-null   float64
 16  REGION             3974

# **Data Analysis**

In [666]:
#Which ones were the most used city for the three days?
df_cells['CITY'].groupby(df_cells['LOCATION']).describe()

count unique           top freq
LOCATION                                 
ABIDJAN    1669     15        COCODY  354
INTERIEUR  2305    490  YAMOUSSOUKRO  104

In [669]:
#Which cities were the most used during business and home hours?
#Office Hours-Mon-Fri 0730-1200 and 1430-1800, Sat 0800-1200.
df_cells_Telcom.between_time(2012-05-06 14:30:00,18:00:00,include_start=True,include_end=True,axis=None)

SyntaxError: ignored

In [674]:
#Most popular product?
df_Telcom.groupby('PRODUCT')['VALUE'].sum()

PRODUCT
Voice    306716
data      17285
sms       52377
Name: VALUE, dtype: int64

Voice is the most used product in the three days.

In [678]:
#Most active cell id?
df_Telcom.groupby('CELL_ID')['VALUE'].max()

CELL_ID
++ctND+jGD      64
+38u2u/rfx     138
+854AcBQT2     100
+Fs4mTvdKx      31
+QZ9qRtNxA      30
              ... 
ff97ccaf78      37
ff99709d62      50
ffa6759bb2    4440
ffb86d66d2      40
ffcc507579      38
Name: VALUE, Length: 2677, dtype: int64

++ctND+jGD is the most active Cell_ID.

In [688]:
#Which cities are in service
df_cells[['CITY','STATUS']]

CITY      STATUS
0       ADJAME  In Service
1       ADJAME  In Service
2       ADJAME  In Service
3      ASSINIE  In Service
4      ASSINIE  In Service
...        ...         ...
3969  ZUENOULA  In Service
3970  ZUENOULA  In Service
3971  ZUENOULA  In Service
3972  ZUENOULA  In Service
3973  ZUENOULA  In Service

[3974 rows x 2 columns]

In [ ]:
#Most used city for the three days

**Recommendations**

Voice is the most popular product and therefore they should improve their infrastructure in that sector.

++ctND+jGD is the most active Cell_ID thus they should improve on the cell.

They should provide infrastructure to ensure that the cities remain in service.

They should also expand their network in Abidjan and Interieur since they have a wide range of customers there. 